### PR TITLE
HistogramData rollout: algorithms 6 to 10

### LINKED ISSUE #17648: HistogramData rollout: algorithms 6 to 10
HistogramData rollout: algorithms <X to Y>

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:
```
Framework/Algorithms/src/PolarizationCorrection.cpp
Framework/Algorithms/src/Q1D2.cpp
Framework/Algorithms/src/Q1DWeighted.cpp
Framework/Algorithms/src/Qhelper.cpp
Framework/Algorithms/src/Qxy.cpp
```

### PR DESCRIPTION
**Pull request by samueljackson92**
## Description of work.

This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

```
Framework/Algorithms/src/PolarizationCorrection.cpp
Framework/Algorithms/src/Q1D2.cpp
Framework/Algorithms/src/Q1DWeighted.cpp
Framework/Algorithms/src/Qhelper.cpp
Framework/Algorithms/src/Qxy.cpp
```

Performance tests were added for:

```
Framework/Algorithms/src/Qxy.cpp
```
## To test.

Code review.

Fixes #17648 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

### CODE REVIEW

File: Framework/Algorithms/src/PolarizationCorrection.cpp [hunk 1]
```diff
@@ -78,9 +78,11 @@ void validateInputWorkspace(WorkspaceGroup_sptr &ws) {
                                       "WorkspaceGroup");
         }
 
-        auto currentX = ws2d->readX(0);
-        auto lastX = lastWS->readX(0);
-        if (currentX != lastX) {
+        auto currentX = ws2d->x(0);
+        auto lastX = lastWS->x(0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/PolarizationCorrection.cpp]: Should this be references? Does not seem to be modified below?

File: Framework/Algorithms/src/Q1D2.cpp [hunk 2]
```diff
@@ -279,21 +279,21 @@ void Q1D2::exec() {
...
-    ws_sumOfNormFactors->dataX(0) = outputWS->dataX(0);
+    ws_sumOfNormFactors->setSharedX(0, outputWS->sharedX(0));
+    ws_sumOfNormFactors->mutableY(0) = normSum;
     ws_sumOfNormFactors->setSharedDx(0, outputWS->sharedDx(0));
-    for (size_t i = 0; i < ws_sumOfNormFactors->dataY(0).size(); i++) {
-      ws_sumOfNormFactors->dataY(0)[i] = normSum[i];
-      ws_sumOfNormFactors->dataE(0)[i] = sqrt(normError2[i]);
-    }
+    std::transform(normError2.cbegin(), normError2.cend(),
+                   ws_sumOfNormFactors->mutableE(0).begin(), sqrt);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Q1D2.cpp]: You could use `ws_sumOfNormFactors->setCountVariances(0, normError2);` here.

File: Framework/Algorithms/src/Q1D2.cpp [hunk 3]
```diff
@@ -279,21 +279,21 @@ void Q1D2::exec() {
...
+    ws_sumOfCounts->setSharedX(0, outputWS->sharedX(0));
+    // Copy now as YOut is modified in normalize
+    ws_sumOfCounts->mutableY(0) = YOut;
     ws_sumOfCounts->setSharedDx(0, outputWS->sharedDx(0));
-    for (size_t i = 0; i < outputWS->dataE(0).size(); i++) {
-      ws_sumOfCounts->dataE(0)[i] = sqrt(outputWS->dataE(0)[i]);
-    }
+    auto outputE = outputWS->e(0);
+    std::transform(outputE.cbegin(), outputE.cend(),
+                   ws_sumOfCounts->mutableE(0).begin(), sqrt);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Q1D2.cpp]: `ws_sumOfCounts->setCountVariances(0, outputWS->e(0));` should work, I think.

File: Framework/Algorithms/src/Q1DWeighted.cpp [hunk 4]
```diff
@@ -295,8 +295,8 @@ void Q1DWeighted::exec() {
         // Normalize wedges
         for (int iWedge = 0; iWedge < nWedges; iWedge++) {
           if (wedge_XNorm[iWedge][k] > 0) {
-            MantidVec &wedgeYOut = wedgeWorkspaces[iWedge]->dataY(0);
-            MantidVec &wedgeEOut = wedgeWorkspaces[iWedge]->dataE(0);
+            auto wedgeYOut = wedgeWorkspaces[iWedge]->mutableY(0);
+            auto wedgeEOut = wedgeWorkspaces[iWedge]->mutableE(0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Q1DWeighted.cpp]: &?

File: Framework/Algorithms/src/Q1DWeighted.cpp [hunk 5]
```diff
@@ -316,8 +316,8 @@ void Q1DWeighted::exec() {
   }
   for (int iWedge = 0; iWedge < nWedges; iWedge++) {
     for (int i = 0; i < sizeOut - 1; i++) {
-      MantidVec &wedgeYOut = wedgeWorkspaces[iWedge]->dataY(0);
-      MantidVec &wedgeEOut = wedgeWorkspaces[iWedge]->dataE(0);
+      auto wedgeYOut = wedgeWorkspaces[iWedge]->mutableY(0);
+      auto wedgeEOut = wedgeWorkspaces[iWedge]->mutableE(0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Q1DWeighted.cpp]: &?

File: Framework/Algorithms/src/Qxy.cpp [hunk 6]
```diff
@@ -381,10 +392,8 @@ Qxy::setUpOutputWorkspace(API::MatrixWorkspace_const_sptr inputWorkspace) {
   // Fill the X vectors in the output workspace
   for (int i = 0; i < bins - 1; ++i) {
     outputWorkspace->setBinEdges(i, axis);
-    for (int j = 0; j < bins - j; ++j) {
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Qxy.cpp]: What is this upper loop bound?? It looks weird, but if it is correct your change below cannot be right.

File: Framework/Algorithms/test/QxyTest.h [hunk 7]
```diff
@@ -138,7 +138,7 @@ class QxyTest : public CxxTest::TestSuite {
         std::cout << i << '\n';
         for (int j = 0 ; j < 30; j ++)
         {
-        std::cout << result->readY(i)[j] << " ";
+        std::cout << result->y(i)[j] << " ";
```
Comment (SimonHeybrock) [inline Framework/Algorithms/test/QxyTest.h]: Maybe just remove the commented code altogether?

**Review by SimonHeybrock — COMMENTED** [review]
Looks good apart from minor comments.

File: Framework/Algorithms/src/PolarizationCorrection.cpp (hunk 1 above)
Comment (samueljackson92) [inline Framework/Algorithms/src/PolarizationCorrection.cpp]: Yes. Done.

File: Framework/Algorithms/test/QxyTest.h (hunk 7 above)
Comment (samueljackson92) [inline Framework/Algorithms/test/QxyTest.h]: Agreed.

File: Framework/Algorithms/src/Q1D2.cpp [hunk 8]
```diff
@@ -279,16 +279,14 @@ void Q1D2::exec() {
     ws_sumOfCounts->mutableY(0) = YOut;
     ws_sumOfCounts->setSharedDx(0, outputWS->sharedDx(0));
     auto outputE = outputWS->e(0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Q1D2.cpp]: Can this line be removed?

File: Framework/Algorithms/src/Qxy.cpp [hunk 9]
```diff
@@ -392,8 +391,12 @@ Qxy::setUpOutputWorkspace(API::MatrixWorkspace_const_sptr inputWorkspace) {
   // Fill the X vectors in the output workspace
   for (int i = 0; i < bins - 1; ++i) {
     outputWorkspace->setBinEdges(i, axis);
-    outputWorkspace->mutableY(i) = std::numeric_limits<double>::quiet_NaN();
-    outputWorkspace->mutableE(i) = std::numeric_limits<double>::quiet_NaN();
+    for (int j = 0; j < bins - j; ++j) {
+      outputWorkspace->mutableY(i)[j] =
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/Qxy.cpp]: Just like the old `dataY` methods `mutableY` is slightly expensive (due to the copy-on-write mechanism), so it should be avoided to call it in a loop for each bin (for each histogram). That is, try to store a reference to it before the loop over `j`.

**Review by SimonHeybrock — COMMENTED** [review]
Just to more small comments, then this should be ready:

File: Framework/Algorithms/src/Q1D2.cpp (hunk 8 above)
Comment (samueljackson92) [inline Framework/Algorithms/src/Q1D2.cpp]: Yes.

File: Framework/Algorithms/src/Qxy.cpp (hunk 9 above)
Comment (samueljackson92) [inline Framework/Algorithms/src/Qxy.cpp]: Sure.

**Comment by SimonHeybrock** [comment]
Looks good now, :shipit: once the builds pass.

**Pull request merged**